vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,787 @@
+
+  n->left = save_load_node(b, std::move(n->left), model_file, read, text, msg);
+  n->right = save_load_node(b, std::move(n->right), model_file, read, text, msg);
+
+  return n;
+}
+
+void save_load_tree(emt_tree& b, io_buf& model_file, bool read, bool text)
+{
+  std::stringstream msg;
+  if (model_file.num_files() > 0)
+  {
+    uint32_t ss = b.all->weights.stride_shift();
+    WRITEIT(ss, "stride_shift");
+
+    b.all->weights.stride_shift(ss);
+
+    int scorer_type = static_cast<int>(b.scorer_type);
+    int router_type = static_cast<int>(b.router_type);
+    uint32_t tree_bound = b.bounder->max_size;
+
+    WRITEIT(b.leaf_split, "leaf_split");
+    WRITEIT(tree_bound, "tree_bound");
+    WRITEIT(scorer_type, "scorer_type");
+    WRITEIT(router_type, "router_type");
+
+    b.scorer_type = static_cast<emt_scorer_type>(scorer_type);
+    b.router_type = static_cast<emt_router_type>(router_type);
+
+    if (read) { b.bounder = VW::make_unique<emt_lru>(tree_bound); }
+
+    b.root = save_load_node(b, std::move(b.root), model_file, read, text, msg);
+    *(b.all->trace_message) << "done loading...." << std::endl;
+  }
+}
+/////////////////End of Save & Load/////////////////////////////////
+////////////////////////////////////////////////////////////////////
+}  // namespace eigen_memory_tree
+}  // namespace reductions
+}  // namespace VW
+
+base_learner* VW::reductions::eigen_memory_tree_setup(VW::setup_base_i& stack_builder)
+{
+  options_i& options = *stack_builder.get_options();
+  VW::workspace& all = *stack_builder.get_all_pointer();
+
+  auto t = VW::make_unique<VW::reductions::eigen_memory_tree::emt_tree>();
+  bool _;
+  uint32_t scorer_type = 0, router_type = 0, tree_bound = 0;
+
+  option_group_definition new_options("[Reduction] Eigen Memory Tree");
+  new_options.add(make_option("eigen_memory_tree", _).keep().necessary().help("Make an eigen memory tree"))
+      .add(make_option("emt_tree", tree_bound)
+               .keep()
+               .default_value(0)
+               .help("Indicates the maximum number of memories the tree can have"))
+      .add(make_option("emt_leaf", t->leaf_split)
+               .keep()
+               .default_value(100)
+               .help("Indicates the maximum number of memories a leaf can have"))
+      .add(make_option("emt_scorer", scorer_type)
+               .keep()
+               .one_of({1, 2, 3, 4})
+               .default_value(3)
+               .help("Indicates the type of scorer to use (1=random;2=distance;3=self consistent rank;4=not self "
+                     "consistent rank)"))
+      .add(make_option("emt_router", router_type)
+               .keep()
+               .one_of({1, 2})
+               .default_value(2)
+               .help("Indicates the type of router to use (1=random;2=eigen)"));
+
+  if (!options.add_parse_and_check_necessary(new_options)) { return nullptr; }
+
+  t->scorer_type = static_cast<VW::reductions::eigen_memory_tree::emt_scorer_type>(scorer_type);
```

Comment:
it looks a bit dangerous (i e if someone adds new scorer_type in the middle). Probably helper conversion from string constant to enum (either with switch/case or using std::map) with explicit error for unknown names is a bit better here

Reply:
@ataymano That's why I explicitly set the values of the enum... Does it still seem dangerous to you?